File: lib/ts/src/internal/types/UUnion.ts
import { ValidationFailure } from "../../internal/validation/ValidationFailure";
import { UStruct } from "../../internal/types/UStruct";
import { UTypeDeclaration } from "../../internal/types/UTypeDeclaration";
import { validateUnion } from "../../internal/validation/ValidateUnion";
import { generateRandomUnion } from "../../internal/generation/GenerateRandomUnion";
import { UType } from "../../internal/types/UType";
import { GenerateContext } from "../../internal/generation/GenerateContext";
import { ValidateContext } from "../validation/ValidateContext";

export const unionName: string = "Object";

export class UUnion implements UType {
    name: string;
    tags: { [key: string]: UStruct };
    tagIndices: { [key: string]: number };

    constructor(name: string, tags: { [key: string]: UStruct }, tagIndices: { [key: string]: number }) {
        this.name = name;
        this.tags = tags;
        this.tagIndices = tagIndices;
    }

    getTypeParameterCount(): number {
        return 0;
    }

    validate(value: any, typeParameters: UTypeDeclaration[], ctx: ValidateContext): ValidationFailure[] {
        return validateUnion(value, this.name, this.tags, ctx);
    }

    generateRandomValue(
        blueprintValue: any,
        useBlueprintValue: boolean,
        typeParameters: UTypeDeclaration[],
        ctx: GenerateContext,
    ): any {
        return generateRandomUnion(blueprintValue, useBlueprintValue, this.tags, ctx);
    }

    getName(): string {
        return unionName;
    }
}

File: tool/uapicodegen/uapicodegen/uapi/internal/types/UUnion.py
from typing import TYPE_CHECKING
from ...internal.types.UType import UType

if TYPE_CHECKING:
    from ...internal.validation.ValidateContext import ValidateContext
    from ...internal.generation.GenerateContext import GenerateContext
    from ...RandomGenerator import RandomGenerator
    from ...internal.validation.ValidationFailure import ValidationFailure
    from ...internal.types.UStruct import UStruct
    from ...internal.types.UTypeDeclaration import UTypeDeclaration

_UNION_NAME: str = "Object"


class UUnion(UType):

    def __init__(self, name: str, tags: dict[str, 'UStruct'], tag_indices: dict[str, int]) -> None:
        self.name = name
        self.tags = tags
        self.tag_indices = tag_indices

    def get_type_parameter_count(self) -> int:
        return 0

    def validate(self, value: object, type_parameters: list['UTypeDeclaration'], ctx: 'ValidateContext') -> list['ValidationFailure']:
        from ...internal.validation.ValidateUnion import validate_union
        return validate_union(value, self.name, self.tags, ctx)

    def generate_random_value(self, blueprint_value: object, use_blueprint_value: bool, type_parameters: list['UTypeDeclaration'], ctx: 'GenerateContext') -> object:
        from ...internal.generation.GenerateRandomUnion import generate_random_union
        return generate_random_union(blueprint_value, use_blueprint_value, self.tags, ctx)

    def get_name(self) -> str:
        return _UNION_NAME
